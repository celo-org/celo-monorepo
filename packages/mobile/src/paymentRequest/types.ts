import { SHORT_CURRENCIES } from 'src/geth/consts'
import { NotificationTypes } from 'src/notifications/types'

export enum PaymentRequestStatus {
  REQUESTED = 'REQUESTED',
  COMPLETED = 'COMPLETED',
  DECLINED = 'DECLINED',
  CANCELLED = 'CANCELLED',
}

export interface PaymentRequest {
  uid?: string
  amount: string
  comment?: string
  timestamp: Date
  requesterAddress: string
  requesteeAddress: string
  requesterE164Number?: string
  currency: SHORT_CURRENCIES
  status: PaymentRequestStatus
  notified: boolean
  type?: NotificationTypes.PAYMENT_REQUESTED
}
