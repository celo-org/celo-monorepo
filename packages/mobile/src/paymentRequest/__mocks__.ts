import { PaymentRequest, PaymentRequestStatus } from 'src/account/types'
import { SHORT_CURRENCIES } from 'src/geth/consts'

export function paymentRequestDouble(partial: object): PaymentRequest {
  return {
    uid: '7363bvby2ba8270273',
    amount: '20',
    comment: 'Just the best',
    requesterE164Number: '+1555-867-5309',
    requesteeAddress: '0x15280126303735b625',
    currency: SHORT_CURRENCIES.DOLLAR,
    timestamp: new Date(),
    requesterAddress: '101929292929',
    status: PaymentRequestStatus.REQUESTED,
    notified: true,
    ...partial,
  }
}
