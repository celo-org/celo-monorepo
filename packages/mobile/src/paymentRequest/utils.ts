import { PaymentRequest } from 'src/account/types'
import { AddressToE164NumberType } from 'src/identity/reducer'
import { NumberToRecipient, Recipient, RecipientKind } from 'src/recipients/recipient'

export function getRecipientFromPaymentRequest(
  paymentRequest: PaymentRequest,
  recipientCache: NumberToRecipient
): Recipient {
  const cachedRecipient = paymentRequest.requesterE164Number
    ? recipientCache[paymentRequest.requesterE164Number]
    : null

  if (cachedRecipient) {
    return {
      ...cachedRecipient,
      kind: RecipientKind.Address,
      address: paymentRequest.requesterAddress,
    }
  } else {
    return {
      kind: RecipientKind.Address,
      address: paymentRequest.requesterAddress,
      displayName: paymentRequest.requesterE164Number || paymentRequest.requesterAddress,
    }
  }
}

export function getSenderFromPaymentRequest(
  paymentRequest: PaymentRequest,
  addressToE164Number: AddressToE164NumberType,
  recipientCache: NumberToRecipient
): Recipient {
  const e164PhoneNumber = addressToE164Number[paymentRequest.requesteeAddress]
  if (!e164PhoneNumber) {
    return {
      kind: RecipientKind.Address,
      address: paymentRequest.requesteeAddress,
      displayName: paymentRequest.requesteeAddress,
    }
  }

  const cachedRecipient = recipientCache[e164PhoneNumber]
  if (cachedRecipient) {
    return {
      ...cachedRecipient,
      kind: RecipientKind.Address,
      address: paymentRequest.requesteeAddress,
    }
  } else {
    return {
      kind: RecipientKind.MobileNumber,
      address: paymentRequest.requesterAddress,
      e164PhoneNumber,
      displayName: e164PhoneNumber,
    }
  }
}
