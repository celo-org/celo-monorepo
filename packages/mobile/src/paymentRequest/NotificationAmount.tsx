import colors from '@celo/react-components/styles/colors'
import fontStyles from '@celo/react-components/styles/fonts'
import * as React from 'react'
import { StyleSheet, Text, View } from 'react-native'
import { CURRENCIES, CURRENCY_ENUM } from 'src/geth/consts'
import { getCentAwareMoneyDisplay } from 'src/utils/formatting'

interface Props {
  amount: string
}

export default class NotificationAmount extends React.PureComponent<Props> {
  render() {
    return (
      <View>
        <Text style={[fontStyles.bodySmallSemiBold, styles.amount]}>
          {CURRENCIES[CURRENCY_ENUM.DOLLAR].symbol + getCentAwareMoneyDisplay(this.props.amount)}
        </Text>
      </View>
    )
  }
}

const styles = StyleSheet.create({
  amount: {
    color: colors.darkSecondary,
  },
})
