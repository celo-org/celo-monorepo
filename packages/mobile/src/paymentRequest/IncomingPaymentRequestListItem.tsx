import BaseNotification from '@celo/react-components/components/BaseNotification'
import ContactCircle from '@celo/react-components/components/ContactCircle'
import fontStyles from '@celo/react-components/styles/fonts'
import BigNumber from 'bignumber.js'
import * as React from 'react'
import { WithTranslation } from 'react-i18next'
import { Image, StyleSheet, Text, View } from 'react-native'
import { PaymentRequestStatus } from 'src/account/types'
import CeloAnalytics from 'src/analytics/CeloAnalytics'
import { CustomEventNames } from 'src/analytics/constants'
import { TokenTransactionType } from 'src/apollo/types'
import { updatePaymentRequestStatus } from 'src/firebase/actions'
import { CURRENCIES, CURRENCY_ENUM } from 'src/geth/consts'
import { Namespaces, withTranslation } from 'src/i18n'
import { unknownUserIcon } from 'src/images/Images'
import { navigate } from 'src/navigator/NavigationService'
import { Screens } from 'src/navigator/Screens'
import { getRecipientThumbnail, Recipient } from 'src/recipients/recipient'
import { getCentAwareMoneyDisplay } from 'src/utils/formatting'
import Logger from 'src/utils/Logger'

interface OwnProps {
  requester: Recipient
  amount: string
  comment: string
  id: string
  updatePaymentRequestStatus: typeof updatePaymentRequestStatus
}

type Props = OwnProps & WithTranslation

const AVATAR_SIZE = 40

export class IncomingPaymentRequestListItem extends React.Component<Props> {
  onPay = () => {
    const { amount, comment: reason, requester: recipient } = this.props
    navigate(Screens.SendConfirmation, {
      confirmationInput: {
        reason,
        recipient,
        amount: new BigNumber(amount),
        recipientAddress: recipient.address,
        type: TokenTransactionType.PayRequest,
      },
      onConfirm: this.onPaymentSuccess,
      onCancel: this.onPaymentDecline,
    })
  }

  onPaymentSuccess = () => {
    const { id } = this.props
    this.props.updatePaymentRequestStatus(id.toString(), PaymentRequestStatus.COMPLETED)
    Logger.showMessage(this.props.t('requestPaid'))
    CeloAnalytics.track(CustomEventNames.incoming_request_payment_pay)
  }

  onPaymentDecline = () => {
    const { id } = this.props
    this.props.updatePaymentRequestStatus(id.toString(), PaymentRequestStatus.DECLINED)
    Logger.showMessage(this.props.t('requestDeclined'))
    CeloAnalytics.track(CustomEventNames.incoming_request_payment_decline)
  }

  getCTA = () => {
    return [
      {
        text: this.props.t('global:pay'),
        onPress: this.onPay,
      },
      {
        text: this.props.t('global:decline'),
        onPress: this.onPaymentDecline,
      },
    ]
  }

  render() {
    const { requester, t } = this.props
    return (
      <View style={styles.container}>
        <BaseNotification
          icon={
            <ContactCircle
              size={AVATAR_SIZE}
              address={requester.address}
              name={requester.displayName}
              thumbnailPath={getRecipientThumbnail(requester)}
            >
              <Image source={unknownUserIcon} style={styles.unknownUser} />
            </ContactCircle>
          }
          title={t('incomingPaymentRequestNotificationTitle', {
            name: requester.displayName,
            amount:
              CURRENCIES[CURRENCY_ENUM.DOLLAR].symbol + getCentAwareMoneyDisplay(this.props.amount),
          })}
          ctas={this.getCTA()}
          onPress={this.onPay}
        >
          <Text style={fontStyles.bodySmall}>{this.props.comment || t('defaultComment')}</Text>
        </BaseNotification>
      </View>
    )
  }
}

const styles = StyleSheet.create({
  container: {
    marginBottom: 16,
  },
  unknownUser: {
    height: AVATAR_SIZE,
    width: AVATAR_SIZE,
    justifyContent: 'center',
    alignItems: 'center',
  },
})

export default withTranslation(Namespaces.paymentRequestFlow)(IncomingPaymentRequestListItem)
