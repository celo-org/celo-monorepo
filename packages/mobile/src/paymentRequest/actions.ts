import { PaymentRequest, PaymentRequestStatus } from 'src/paymentRequest/types'

export enum Actions {
  WRITE_PAYMENT_REQUEST = 'PAYMENT_REQUEST/WRITE_PAYMENT_REQUEST',
  UPDATE_INCOMING_REQUESTS = 'PAYMENT_REQUEST/UPDATE_INCOMING_REQUESTS',
  UPDATE_OUTGOING_REQUESTS = 'PAYMENT_REQUEST/UPDATE_OUTGOING_REQUESTS',
  UPDATE_REQUEST_STATUS = 'PAYMENT_REQUEST/UPDATE_REQUEST_STATUS',
  UPDATE_REQUEST_NOTIFIED = 'PAYMENT_REQUEST/UPDATE_REQUEST_NOTIFIED',
}

export interface WritePaymentRequestAction {
  type: Actions.WRITE_PAYMENT_REQUEST
  paymentRequest: PaymentRequest
}

export const writePaymentRequest = (paymentRequest: PaymentRequest): WritePaymentRequestAction => ({
  type: Actions.WRITE_PAYMENT_REQUEST,
  paymentRequest,
})

export interface UpdateIncomingPaymentRequestsAction {
  type: Actions.UPDATE_INCOMING_REQUESTS
  paymentRequests: PaymentRequest[]
}

export const updateIncomingPaymentRequests = (
  paymentRequests: PaymentRequest[]
): UpdateIncomingPaymentRequestsAction => ({
  type: Actions.UPDATE_INCOMING_REQUESTS,
  paymentRequests,
})

export interface UpdateOutgoingPaymentRequestsAction {
  type: Actions.UPDATE_OUTGOING_REQUESTS
  paymentRequests: PaymentRequest[]
}

export const updateOutgoingPaymentRequests = (
  paymentRequests: PaymentRequest[]
): UpdateOutgoingPaymentRequestsAction => ({
  type: Actions.UPDATE_OUTGOING_REQUESTS,
  paymentRequests,
})

export interface UpdatePaymentRequestNotifiedAction {
  type: Actions.UPDATE_REQUEST_NOTIFIED
  id: string
  notified: boolean
}

export const updatePaymentRequestNotified = (
  id: string,
  notified: boolean
): UpdatePaymentRequestNotifiedAction => ({
  type: Actions.UPDATE_REQUEST_NOTIFIED,
  id,
  notified,
})

export interface CompletePaymentRequestAction {
  type: Actions.UPDATE_REQUEST_STATUS
  status: PaymentRequestStatus.COMPLETED
  id: string
}

export const completePaymentRequest = (id: string): CompletePaymentRequestAction => ({
  type: Actions.UPDATE_REQUEST_STATUS,
  status: PaymentRequestStatus.COMPLETED,
  id,
})

export interface DeclinePaymentRequestAction {
  type: Actions.UPDATE_REQUEST_STATUS
  status: PaymentRequestStatus.DECLINED
  id: string
}

export const declinePaymentRequest = (id: string): DeclinePaymentRequestAction => ({
  type: Actions.UPDATE_REQUEST_STATUS,
  status: PaymentRequestStatus.DECLINED,
  id,
})

export interface CancelPaymentRequestAction {
  type: Actions.UPDATE_REQUEST_STATUS
  status: PaymentRequestStatus.CANCELLED
  id: string
}

export const cancelPaymentRequest = (id: string): CancelPaymentRequestAction => ({
  type: Actions.UPDATE_REQUEST_STATUS,
  status: PaymentRequestStatus.CANCELLED,
  id,
})

export type ActionTypes = UpdateIncomingPaymentRequestsAction | UpdateOutgoingPaymentRequestsAction
