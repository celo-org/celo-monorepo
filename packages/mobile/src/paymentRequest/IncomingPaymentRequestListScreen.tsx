import React from 'react'
import { WithNamespaces, withNamespaces } from 'react-i18next'
import { NavigationInjectedProps } from 'react-navigation'
import { connect } from 'react-redux'
import { getIncomingPaymentRequests } from 'src/account/selectors'
import { PaymentRequest } from 'src/account/types'
import { updatePaymentRequestStatus } from 'src/firebase/actions'
import i18n, { Namespaces } from 'src/i18n'
import { fetchPhoneAddresses } from 'src/identity/actions'
import { e164NumberToAddressSelector, E164NumberToAddressType } from 'src/identity/reducer'
import {
  NotificationList,
  titleWithBalanceNavigationOptions,
  useBalanceInNavigationParam,
} from 'src/notifications/NotificationList'
import IncomingPaymentRequestListItem from 'src/paymentRequest/IncomingPaymentRequestListItem'
import { getRecipientFromPaymentRequest } from 'src/paymentRequest/utils'
import { NumberToRecipient } from 'src/recipients/recipient'
import { recipientCacheSelector } from 'src/recipients/reducer'
import { RootState } from 'src/redux/reducers'

interface StateProps {
  dollarBalance: string | null
  paymentRequests: PaymentRequest[]
  e164PhoneNumberAddressMapping: E164NumberToAddressType
  recipientCache: NumberToRecipient
}

interface DispatchProps {
  updatePaymentRequestStatus: typeof updatePaymentRequestStatus
  fetchPhoneAddresses: typeof fetchPhoneAddresses
}

const mapStateToProps = (state: RootState): StateProps => ({
  dollarBalance: state.stableToken.balance,
  paymentRequests: getIncomingPaymentRequests(state),
  e164PhoneNumberAddressMapping: e164NumberToAddressSelector(state),
  recipientCache: recipientCacheSelector(state),
})

type Props = NavigationInjectedProps & WithNamespaces & StateProps & DispatchProps

export const listItemRenderer = (params: {
  recipientCache: NumberToRecipient
  updatePaymentRequestStatus: typeof updatePaymentRequestStatus
}) => (request: PaymentRequest, key: number | undefined = undefined) => {
  const requester = getRecipientFromPaymentRequest(request, params.recipientCache)

  return (
    <IncomingPaymentRequestListItem
      key={key}
      id={request.uid || ''}
      amount={request.amount}
      updatePaymentRequestStatus={params.updatePaymentRequestStatus}
      requester={requester}
      comment={request.comment}
    />
  )
}

const IncomingPaymentRequestListScreen = (props: Props) => {
  const { recipientCache, dollarBalance, navigation } = props
  useBalanceInNavigationParam(dollarBalance, navigation)
  return (
    <NotificationList
      items={props.paymentRequests}
      listItemRenderer={listItemRenderer({
        updatePaymentRequestStatus: props.updatePaymentRequestStatus,
        recipientCache,
      })}
      dollarBalance={props.dollarBalance}
    />
  )
}

IncomingPaymentRequestListScreen.navigationOptions = titleWithBalanceNavigationOptions(
  i18n.t('walletFlow5:incomingPaymentRequests')
)

export default connect<StateProps, DispatchProps, {}, RootState>(
  mapStateToProps,
  {
    updatePaymentRequestStatus,
    fetchPhoneAddresses,
  }
)(withNamespaces(Namespaces.paymentRequestFlow)(IncomingPaymentRequestListScreen))
