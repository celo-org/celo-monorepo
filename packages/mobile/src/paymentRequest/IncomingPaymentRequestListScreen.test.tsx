import * as React from 'react'
import 'react-native'
import { Provider } from 'react-redux'
import * as renderer from 'react-test-renderer'
import { PaymentRequest } from 'src/account/types'
import { paymentRequestDouble } from 'src/paymentRequest/__mocks__'
import IncomingPaymentRequestListScreen from 'src/paymentRequest/IncomingPaymentRequestListScreen'
import { createMockStore } from 'test/utils'
import { mockAccount, mockE164Number } from 'test/values'

const requests = [
  paymentRequestDouble({
    amount: '20',
    comment: 'Just the best',
    requesterE164Number: '+1555-867-5309',
    requesterAddress: mockAccount,
  }),
  paymentRequestDouble({
    amount: '102',
    comment: 'Just the best for the best. Thanos & Zeus Gods of ultimate Power',
    requesterE164Number: mockE164Number,
    requesterAddress: mockAccount,
  }),
  paymentRequestDouble({
    amount: '1',
    comment: 'Just the best but less',
    requesterE164Number: mockE164Number,
    requesterAddress: mockAccount,
  }),
]

function testStore(incomingPaymentRequests: PaymentRequest[]) {
  return createMockStore({
    stableToken: { balance: '120' },
    account: { incomingPaymentRequests },
  })
}

describe('IncomingPaymentRequestListScreen', () => {
  it('renders correctly with requests', () => {
    const store = testStore(requests)

    const tree = renderer.create(
      <Provider store={store}>
        <IncomingPaymentRequestListScreen />
      </Provider>
    )
    expect(tree).toMatchSnapshot()
  })

  it('renders correctly with no requests', () => {
    const store = testStore([])

    const tree = renderer.create(
      <Provider store={store}>
        <IncomingPaymentRequestListScreen />
      </Provider>
    )
    expect(tree).toMatchSnapshot()
  })
})
