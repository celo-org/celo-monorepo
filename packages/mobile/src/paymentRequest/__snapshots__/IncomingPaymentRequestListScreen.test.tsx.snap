// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`IncomingPaymentRequestListScreen renders correctly with no requests 1`] = `
<SafeAreaView
  style={
    Object {
      "backgroundColor": "#FFFFFF",
      "flex": 1,
    }
  }
>
  <Text
    style={
      Array [
        Object {
          "color": "#81868B",
          "fontFamily": "Hind-Regular",
          "fontSize": 16,
          "lineHeight": 24,
        },
        Object {
          "marginTop": 30,
          "textAlign": "center",
        },
      ]
    }
  >
    global:emptyList
  </Text>
</SafeAreaView>
`;

exports[`IncomingPaymentRequestListScreen renders correctly with requests 1`] = `
<SafeAreaView
  style={
    Object {
      "backgroundColor": "#FFFFFF",
      "flex": 1,
    }
  }
>
  <RCTScrollView>
    <View>
      <View
        style={
          Object {
            "margin": 16,
          }
        }
      >
        <View
          style={
            Object {
              "marginBottom": 16,
            }
          }
        >
          <View
            accessible={true}
            focusable={true}
            isTVSelectable={true}
            nativeBackgroundAndroid={
              Object {
                "attribute": "selectableItemBackground",
                "type": "ThemeAttrAndroid",
              }
            }
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
          >
            <View
              style={
                Array [
                  Object {
                    "backgroundColor": "#FFFFFF",
                    "flexDirection": "row",
                    "justifyContent": "space-between",
                    "padding": 16,
                    "width": "100%",
                  },
                  Object {
                    "elevation": 2,
                    "shadowColor": "black",
                    "shadowOffset": Object {
                      "height": 1,
                      "width": 0,
                    },
                    "shadowOpacity": 0.3,
                    "shadowRadius": 1.6,
                  },
                ]
              }
            >
              <View
                style={
                  Object {
                    "alignItems": "center",
                    "paddingRight": 16,
                  }
                }
              >
                <View
                  style={
                    Array [
                      Object {
                        "alignItems": "center",
                        "flexDirection": "row",
                        "justifyContent": "center",
                      },
                      undefined,
                    ]
                  }
                >
                  <View
                    style={
                      Array [
                        Object {
                          "alignItems": "center",
                          "justifyContent": "center",
                        },
                        Object {
                          "backgroundColor": "#7AD6FE",
                          "borderRadius": 20,
                          "height": 40,
                          "width": 40,
                        },
                      ]
                    }
                  >
                    <Image
                      source={
                        Object {
                          "testUri": "../../../packages/mobile/src/images/unknown-user-icon.png",
                        }
                      }
                      style={
                        Object {
                          "alignItems": "center",
                          "height": 40,
                          "justifyContent": "center",
                          "width": 40,
                        }
                      }
                    />
                  </View>
                </View>
              </View>
              <View
                style={
                  Object {
                    "flex": 1,
                    "justifyContent": "space-between",
                  }
                }
              >
                <Text
                  style={
                    Object {
                      "color": "#81868B",
                      "fontFamily": "Hind-Regular",
                      "fontSize": 14,
                      "lineHeight": 18,
                    }
                  }
                >
                  incomingPaymentRequestNotificationTitle
                </Text>
                <View
                  style={
                    Object {
                      "justifyContent": "space-between",
                      "minHeight": 60,
                      "paddingTop": 8,
                    }
                  }
                >
                  <Text
                    style={
                      Object {
                        "color": "#2E3338",
                        "fontFamily": "Hind-Regular",
                        "fontSize": 14,
                        "lineHeight": 18,
                      }
                    }
                  >
                    Just the best
                  </Text>
                  <View
                    style={
                      Object {
                        "flexDirection": "row",
                        "marginTop": 5,
                      }
                    }
                  >
                    <View
                      accessible={true}
                      focusable={true}
                      isTVSelectable={true}
                      nativeBackgroundAndroid={
                        Object {
                          "attribute": "selectableItemBackgroundBorderless",
                          "type": "ThemeAttrAndroid",
                        }
                      }
                      onClick={[Function]}
                      onResponderGrant={[Function]}
                      onResponderMove={[Function]}
                      onResponderRelease={[Function]}
                      onResponderTerminate={[Function]}
                      onResponderTerminationRequest={[Function]}
                      onStartShouldSetResponder={[Function]}
                    >
                      <Text
                        style={
                          Array [
                            Object {
                              "color": "#42D689",
                              "fontFamily": "Hind-SemiBold",
                              "fontSize": 14,
                              "lineHeight": 18,
                            },
                            Object {
                              "paddingEnd": 15,
                            },
                          ]
                        }
                      >
                        global:pay
                      </Text>
                    </View>
                    <View
                      accessible={true}
                      focusable={true}
                      isTVSelectable={true}
                      nativeBackgroundAndroid={
                        Object {
                          "attribute": "selectableItemBackgroundBorderless",
                          "type": "ThemeAttrAndroid",
                        }
                      }
                      onClick={[Function]}
                      onResponderGrant={[Function]}
                      onResponderMove={[Function]}
                      onResponderRelease={[Function]}
                      onResponderTerminate={[Function]}
                      onResponderTerminationRequest={[Function]}
                      onStartShouldSetResponder={[Function]}
                    >
                      <Text
                        style={
                          Array [
                            Object {
                              "color": "#42D689",
                              "fontFamily": "Hind-SemiBold",
                              "fontSize": 14,
                              "lineHeight": 18,
                            },
                            Object {
                              "paddingEnd": 15,
                            },
                          ]
                        }
                      >
                        global:decline
                      </Text>
                    </View>
                  </View>
                </View>
              </View>
            </View>
          </View>
        </View>
        <View
          style={
            Object {
              "marginBottom": 16,
            }
          }
        >
          <View
            accessible={true}
            focusable={true}
            isTVSelectable={true}
            nativeBackgroundAndroid={
              Object {
                "attribute": "selectableItemBackground",
                "type": "ThemeAttrAndroid",
              }
            }
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
          >
            <View
              style={
                Array [
                  Object {
                    "backgroundColor": "#FFFFFF",
                    "flexDirection": "row",
                    "justifyContent": "space-between",
                    "padding": 16,
                    "width": "100%",
                  },
                  Object {
                    "elevation": 2,
                    "shadowColor": "black",
                    "shadowOffset": Object {
                      "height": 1,
                      "width": 0,
                    },
                    "shadowOpacity": 0.3,
                    "shadowRadius": 1.6,
                  },
                ]
              }
            >
              <View
                style={
                  Object {
                    "alignItems": "center",
                    "paddingRight": 16,
                  }
                }
              >
                <View
                  style={
                    Array [
                      Object {
                        "alignItems": "center",
                        "flexDirection": "row",
                        "justifyContent": "center",
                      },
                      undefined,
                    ]
                  }
                >
                  <View
                    style={
                      Array [
                        Object {
                          "alignItems": "center",
                          "justifyContent": "center",
                        },
                        Object {
                          "backgroundColor": "#7AD6FE",
                          "borderRadius": 20,
                          "height": 40,
                          "width": 40,
                        },
                      ]
                    }
                  >
                    <Image
                      source={
                        Object {
                          "testUri": "../../../packages/mobile/src/images/unknown-user-icon.png",
                        }
                      }
                      style={
                        Object {
                          "alignItems": "center",
                          "height": 40,
                          "justifyContent": "center",
                          "width": 40,
                        }
                      }
                    />
                  </View>
                </View>
              </View>
              <View
                style={
                  Object {
                    "flex": 1,
                    "justifyContent": "space-between",
                  }
                }
              >
                <Text
                  style={
                    Object {
                      "color": "#81868B",
                      "fontFamily": "Hind-Regular",
                      "fontSize": 14,
                      "lineHeight": 18,
                    }
                  }
                >
                  incomingPaymentRequestNotificationTitle
                </Text>
                <View
                  style={
                    Object {
                      "justifyContent": "space-between",
                      "minHeight": 60,
                      "paddingTop": 8,
                    }
                  }
                >
                  <Text
                    style={
                      Object {
                        "color": "#2E3338",
                        "fontFamily": "Hind-Regular",
                        "fontSize": 14,
                        "lineHeight": 18,
                      }
                    }
                  >
                    Just the best for the best. Thanos & Zeus Gods of ultimate Power
                  </Text>
                  <View
                    style={
                      Object {
                        "flexDirection": "row",
                        "marginTop": 5,
                      }
                    }
                  >
                    <View
                      accessible={true}
                      focusable={true}
                      isTVSelectable={true}
                      nativeBackgroundAndroid={
                        Object {
                          "attribute": "selectableItemBackgroundBorderless",
                          "type": "ThemeAttrAndroid",
                        }
                      }
                      onClick={[Function]}
                      onResponderGrant={[Function]}
                      onResponderMove={[Function]}
                      onResponderRelease={[Function]}
                      onResponderTerminate={[Function]}
                      onResponderTerminationRequest={[Function]}
                      onStartShouldSetResponder={[Function]}
                    >
                      <Text
                        style={
                          Array [
                            Object {
                              "color": "#42D689",
                              "fontFamily": "Hind-SemiBold",
                              "fontSize": 14,
                              "lineHeight": 18,
                            },
                            Object {
                              "paddingEnd": 15,
                            },
                          ]
                        }
                      >
                        global:pay
                      </Text>
                    </View>
                    <View
                      accessible={true}
                      focusable={true}
                      isTVSelectable={true}
                      nativeBackgroundAndroid={
                        Object {
                          "attribute": "selectableItemBackgroundBorderless",
                          "type": "ThemeAttrAndroid",
                        }
                      }
                      onClick={[Function]}
                      onResponderGrant={[Function]}
                      onResponderMove={[Function]}
                      onResponderRelease={[Function]}
                      onResponderTerminate={[Function]}
                      onResponderTerminationRequest={[Function]}
                      onStartShouldSetResponder={[Function]}
                    >
                      <Text
                        style={
                          Array [
                            Object {
                              "color": "#42D689",
                              "fontFamily": "Hind-SemiBold",
                              "fontSize": 14,
                              "lineHeight": 18,
                            },
                            Object {
                              "paddingEnd": 15,
                            },
                          ]
                        }
                      >
                        global:decline
                      </Text>
                    </View>
                  </View>
                </View>
              </View>
            </View>
          </View>
        </View>
        <View
          style={
            Object {
              "marginBottom": 16,
            }
          }
        >
          <View
            accessible={true}
            focusable={true}
            isTVSelectable={true}
            nativeBackgroundAndroid={
              Object {
                "attribute": "selectableItemBackground",
                "type": "ThemeAttrAndroid",
              }
            }
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
          >
            <View
              style={
                Array [
                  Object {
                    "backgroundColor": "#FFFFFF",
                    "flexDirection": "row",
                    "justifyContent": "space-between",
                    "padding": 16,
                    "width": "100%",
                  },
                  Object {
                    "elevation": 2,
                    "shadowColor": "black",
                    "shadowOffset": Object {
                      "height": 1,
                      "width": 0,
                    },
                    "shadowOpacity": 0.3,
                    "shadowRadius": 1.6,
                  },
                ]
              }
            >
              <View
                style={
                  Object {
                    "alignItems": "center",
                    "paddingRight": 16,
                  }
                }
              >
                <View
                  style={
                    Array [
                      Object {
                        "alignItems": "center",
                        "flexDirection": "row",
                        "justifyContent": "center",
                      },
                      undefined,
                    ]
                  }
                >
                  <View
                    style={
                      Array [
                        Object {
                          "alignItems": "center",
                          "justifyContent": "center",
                        },
                        Object {
                          "backgroundColor": "#7AD6FE",
                          "borderRadius": 20,
                          "height": 40,
                          "width": 40,
                        },
                      ]
                    }
                  >
                    <Image
                      source={
                        Object {
                          "testUri": "../../../packages/mobile/src/images/unknown-user-icon.png",
                        }
                      }
                      style={
                        Object {
                          "alignItems": "center",
                          "height": 40,
                          "justifyContent": "center",
                          "width": 40,
                        }
                      }
                    />
                  </View>
                </View>
              </View>
              <View
                style={
                  Object {
                    "flex": 1,
                    "justifyContent": "space-between",
                  }
                }
              >
                <Text
                  style={
                    Object {
                      "color": "#81868B",
                      "fontFamily": "Hind-Regular",
                      "fontSize": 14,
                      "lineHeight": 18,
                    }
                  }
                >
                  incomingPaymentRequestNotificationTitle
                </Text>
                <View
                  style={
                    Object {
                      "justifyContent": "space-between",
                      "minHeight": 60,
                      "paddingTop": 8,
                    }
                  }
                >
                  <Text
                    style={
                      Object {
                        "color": "#2E3338",
                        "fontFamily": "Hind-Regular",
                        "fontSize": 14,
                        "lineHeight": 18,
                      }
                    }
                  >
                    Just the best but less
                  </Text>
                  <View
                    style={
                      Object {
                        "flexDirection": "row",
                        "marginTop": 5,
                      }
                    }
                  >
                    <View
                      accessible={true}
                      focusable={true}
                      isTVSelectable={true}
                      nativeBackgroundAndroid={
                        Object {
                          "attribute": "selectableItemBackgroundBorderless",
                          "type": "ThemeAttrAndroid",
                        }
                      }
                      onClick={[Function]}
                      onResponderGrant={[Function]}
                      onResponderMove={[Function]}
                      onResponderRelease={[Function]}
                      onResponderTerminate={[Function]}
                      onResponderTerminationRequest={[Function]}
                      onStartShouldSetResponder={[Function]}
                    >
                      <Text
                        style={
                          Array [
                            Object {
                              "color": "#42D689",
                              "fontFamily": "Hind-SemiBold",
                              "fontSize": 14,
                              "lineHeight": 18,
                            },
                            Object {
                              "paddingEnd": 15,
                            },
                          ]
                        }
                      >
                        global:pay
                      </Text>
                    </View>
                    <View
                      accessible={true}
                      focusable={true}
                      isTVSelectable={true}
                      nativeBackgroundAndroid={
                        Object {
                          "attribute": "selectableItemBackgroundBorderless",
                          "type": "ThemeAttrAndroid",
                        }
                      }
                      onClick={[Function]}
                      onResponderGrant={[Function]}
                      onResponderMove={[Function]}
                      onResponderRelease={[Function]}
                      onResponderTerminate={[Function]}
                      onResponderTerminationRequest={[Function]}
                      onStartShouldSetResponder={[Function]}
                    >
                      <Text
                        style={
                          Array [
                            Object {
                              "color": "#42D689",
                              "fontFamily": "Hind-SemiBold",
                              "fontSize": 14,
                              "lineHeight": 18,
                            },
                            Object {
                              "paddingEnd": 15,
                            },
                          ]
                        }
                      >
                        global:decline
                      </Text>
                    </View>
                  </View>
                </View>
              </View>
            </View>
          </View>
        </View>
      </View>
    </View>
  </RCTScrollView>
</SafeAreaView>
`;
